examples/connect.rs

```diff
@@ -78,60 +97,111 @@ impl fmt::Display for PeerId {
+        });
     }
 
-    let our_conn_info =
-        unwrap!(
-        event_loop.run(service.prepare_connection_info()),
-        "Failed to prepare connection info",
-    );
-    let pub_conn_info = our_conn_info.to_pub_connection_info();
-    println!(
-        "Public connection information:\n{}\n",
-        unwrap!(serde_json::to_string(&pub_conn_info))
-    );
-
-    println!("Enter remote peer public connection info:");
-    let their_info = readln();
-    let their_info: PubConnectionInfo<PeerId> = unwrap!(serde_json::from_str(&their_info));
-
-    let peer =
-        unwrap!(
-        event_loop.run(service.connect(our_conn_info, their_info)),
-        "Failed to connect to given peer",
-    );
-    println!("Connected to peer: {}", peer.uid());
-
-    // Run event loop forever.
-    let res = event_loop.run(empty::<(), ()>());
-    unwrap!(res);
+    let mut core = unwrap!(Core::new());
+    let handle = core.handle();
+
+    let our_uid = rand::random();
+
+    let future = {
+        Service::with_config(&handle, config, our_uid)
+        .map_err(|e| panic!("error starting service: {}", e))
+        .and_then(|service| {
+            service
+            .prepare_connection_info()
+            .map_err(|e| panic!("error preparing connection info: {}", e))
+            .and_then(move |our_priv_info| {
+                let our_pub_info = our_priv_info.to_pub_connection_info();
+                let as_str = unwrap!(serde_json::to_string(&our_pub_info));
+                println!("Our connection info:");
+                println!("{}", as_str);
+                println!("");
+                println!("Copy this info and share it with your connecting partner. Then paste \
+                          their info below.");
+                read_line()
+                .and_then(move |line| {
+                    let their_pub_info: PubConnectionInfo<PeerId> = unwrap!(serde_json::from_str(&line));
+
+                    service
+                    .connect(our_priv_info, their_pub_info)
+                    .map_err(|e| panic!("error connecting to peer: {}", e))
+                    .and_then(move |peer| {
+                        let (peer_sink, peer_stream) = peer.split();
+
+                        println!("You are now connected! say hello :)");
+                        let writer = {
+                            future::loop_fn(peer_sink, |peer_sink| {
+                                read_line()
+                                .and_then(|line| {
+                                    let line = line.into_bytes();
+                                    peer_sink
+                                    .send((0, line))
+                                    .map(|peer_sink| Loop::Continue(peer_sink))
+                                    .map_err(|e| panic!("error sending message to peer: {}", e))
+                                })
+                            })
+                        };
+                        let reader = {
```

Comment:
That's a lot of nesting levels. Would you agree with me to extract some code to functions?

Reply:
I've factored out the part which is irrelevant to making a connection into the `have_a_conversation` function. This also simplifies the example as a whole.